Handling "file exists" errors

When a recording is interrupted (process exiting), the files will probably be left behind. pyCA will then try to start the capture again, failing because the target directory already exists. This is not the best thing because then the rest of this event will not be captured as well, even though the problem might have been only very temporarily.

Possible solutions:
a) move the existing directory
b) append a increasing number to the directory name

The recording should probably end with a `RECORDING_FAILED` state nevertheless, so the user can see that something went wrong. Manual intervention to concat the remaining files will probably be necessary anyway.

Example of this:

````
[pyca.schedule:105:control_loop()] [INFO] Next scheduled recording: 2017-10-02 09:00:00
[pyca.capture:43:start_capture()] [INFO] Start recording
[pyca.capture:108:recording_command()] [INFO] ffmpeg -rtsp_transport tcp -i rtsp://james3a-cam01.uct.ac.za/axis-media/media.amp -t 3000 -pix_fmt yuv420p -flags +global_header -c copy ./recordings/recording-1506927600-3288651838695672832/recording-1506927600-3288651838695672832.mp4
[pyca.capture:82:safe_start_capture()] [ERROR] Recording failed
[pyca.capture:83:safe_start_capture()] [ERROR] Traceback (most recent call last):
  File "pyca/capture.py", line 80, in safe_start_capture
    start_capture(event)
  File "pyca/capture.py", line 65, in start_capture
    tracks = recording_command(event)
  File "pyca/capture.py", line 137, in recording_command
    raise RuntimeError('Recording failed (%i)' % captureproc.returncode)
RuntimeError: Recording failed (1)

[pyca.capture:43:start_capture()] [INFO] Start recording
[pyca.capture:82:safe_start_capture()] [ERROR] Recording failed
[pyca.capture:83:safe_start_capture()] [ERROR] Traceback (most recent call last):
  File "pyca/capture.py", line 80, in safe_start_capture
    start_capture(event)
  File "pyca/capture.py", line 57, in start_capture
    os.mkdir(event.directory())
OSError: [Errno 17] File exists: './recordings/recording-1506927600-3288651838695672832'

[pyca.capture:43:start_capture()] [INFO] Start recording
[pyca.capture:82:safe_start_capture()] [ERROR] Recording failed
[pyca.capture:83:safe_start_capture()] [ERROR] Traceback (most recent call last):
  File "pyca/capture.py", line 80, in safe_start_capture
    start_capture(event)
  File "pyca/capture.py", line 57, in start_capture
    os.mkdir(event.directory())
OSError: [Errno 17] File exists: './recordings/recording-1506927600-3288651838695672832'

[pyca.capture:43:start_capture()] [INFO] Start recording
[pyca.capture:82:safe_start_capture()] [ERROR] Recording failed
[pyca.capture:83:safe_start_capture()] [ERROR] Traceback (most recent call last):
  File "pyca/capture.py", line 80, in safe_start_capture
    start_capture(event)
  File "pyca/capture.py", line 57, in start_capture
    os.mkdir(event.directory())
OSError: [Errno 17] File exists: './recordings/recording-1506927600-3288651838695672832'

[pyca.capture:43:start_capture()] [INFO] Start recording
[pyca.capture:82:safe_start_capture()] [ERROR] Recording failed
[pyca.capture:83:safe_start_capture()] [ERROR] Traceback (most recent call last):
  File "pyca/capture.py", line 80, in safe_start_capture
    start_capture(event)
  File "pyca/capture.py", line 57, in start_capture
    os.mkdir(event.directory())
OSError: [Errno 17] File exists: './recordings/recording-1506927600-3288651838695672832'
````

--- a/etc/pyca.conf
+++ b/etc/pyca.conf
@@ -78,7 +78,7 @@ directory        = './recordings'
 # Default: ffmpeg -nostats -re -f lavfi -r 25 -i testsrc -t {{time}} {{dir}}/{{name}}.webm'
 command          = 'ffmpeg -nostats -re -f lavfi -r 25 -i testsrc -f lavfi -i sine -t {{time}} {{dir}}/{{name}}.webm'
 
-# Flavors of putput files produced by the capture command. One flavors should
+# Flavors of output files produced by the capture command. One flavors should
 # be specified for every output file.
 # Type: list of strings (write as '...', '...')
 # Default: 'presenter/source'

--- a/pyca/capture.py
+++ b/pyca/capture.py
@@ -13,6 +13,7 @@ from pyca.utils import recording_state, update_event_status
 from pyca.config import config
 from pyca.db import get_session, RecordedEvent, UpcomingEvent, Status,\
                     Service, ServiceStatus
+import glob
 import logging
 import os
 import os.path
@@ -56,7 +57,7 @@ def start_capture(upcoming_event):
         db.commit()
 
     try_mkdir(config()['capture']['directory'])
-    os.mkdir(event.directory())
+    try_mkdir(event.directory())
 
     # Set state
     update_event_status(event, Status.RECORDING)
@@ -64,12 +65,17 @@ def start_capture(upcoming_event):
     set_service_status_immediate(Service.CAPTURE, ServiceStatus.BUSY)
 
     # Recording
-    tracks = recording_command(event)
-    event.set_tracks(tracks)
+    files = recording_command(event)
+    # [(flavor,path),…]
+    event.set_tracks(list(zip(config('capture')['flavors'], files)))
     db.commit()
 
     # Set status
-    update_event_status(event, Status.FINISHED_RECORDING)
+    # If part files exist, its an partial recording
+    p = any([glob.glob(f'{f}-part-*') for f in files])
+    state = Status.PARTIAL_RECORDING if p else Status.FINISHED_RECORDING
+    logger.info("Set %s to %s", event.uid, Status.str(state))
+    update_event_status(event, state)
     recording_state(event.uid, 'capture_finished')
     set_service_status_immediate(Service.CAPTURE, ServiceStatus.IDLE)
 
@@ -101,6 +107,25 @@ def recording_command(event):
     cmd = cmd.replace('{{dir}}', event.directory())
     cmd = cmd.replace('{{name}}', event.name())
     cmd = cmd.replace('{{previewdir}}', conf['preview_dir'])
+
+    # Parse files into list
+    files = (f.replace('{{dir}}', event.directory()) for f in conf['files'])
+    files = [f.replace('{{name}}', event.name()) for f in files]
+
+    # Move existing files from previous failed recordings
+    for f in files:
+        if not os.path.exists(f):
+            continue
+        # New filename
+        i = 0
+        while True:
+            new_filename = f'{f}-part-{i}'
+            if not os.path.exists(new_filename):
+                break
+            i += 1
+        # Move file
+        os.rename(f, new_filename)
+        logger.warning("Moved file %s to %s to keep it", f, new_filename)
 
     # Signal configuration
     sigterm_time = conf['sigterm_time']
@@ -153,10 +178,8 @@ def recording_command(event):
     # Reset systemd status
     notify.notify('STATUS=Waiting')
 
-    # Return [(flavor,path),…]
-    files = (f.replace('{{dir}}', event.directory()) for f in conf['files'])
-    files = (f.replace('{{name}}', event.name()) for f in files)
-    return list(zip(conf['flavors'], files))
+    # files
+    return files
 
 
 def control_loop():

--- a/pyca/db.py
+++ b/pyca/db.py
@@ -82,6 +82,7 @@ class Status(Constants):
     UPLOADING = 5
     FAILED_UPLOADING = 6
     FINISHED_UPLOADING = 7
+    PARTIAL_RECORDING = 8
 
 
 class ServiceStatus(Constants):

